community/index/src/main/java/org/neo4j/index/internal/gbptree/CrashGenCleaner.java

```diff
@@ -0,0 +1,245 @@
+    private final long unstableGeneration;
+    private final Monitor monitor;
+    private final long maxKeyCount;
+
+    public CrashGenCleaner( PagedFile pagedFile, TreeNode<?,?> treeNode, long lowTreeNodeId, long highTreeNodeId,
+            long stableGeneration, long unstableGeneration, Monitor monitor )
+    {
+        this.pagedFile = pagedFile;
+        this.treeNode = treeNode;
+        this.lowTreeNodeId = lowTreeNodeId;
+        this.highTreeNodeId = highTreeNodeId;
+        this.stableGeneration = stableGeneration;
+        this.unstableGeneration = unstableGeneration;
+        this.monitor = monitor;
+        this.maxKeyCount = treeNode.internalMaxKeyCount();
+    }
+
+    // === Methods about the execution and threading ===
+
+    public void clean() throws IOException
+    {
+        assert unstableGeneration > stableGeneration;
+        assert unstableGeneration - stableGeneration > 1;
+
+        long startTime = currentTimeMillis();
+        int threads = Runtime.getRuntime().availableProcessors();
+        ExecutorService executor = Executors.newFixedThreadPool( threads );
+        AtomicLong nextId = new AtomicLong( lowTreeNodeId);
+        AtomicReference<Throwable> error = new AtomicReference<>();
+        AtomicInteger cleanedPointers = new AtomicInteger();
+        for ( int i = 0; i < threads; i++ )
+        {
+            executor.submit( cleaner( nextId, error, cleanedPointers ) );
+        }
+        executor.shutdown();
+
+        try
+        {
+            long lastProgression = nextId.get();
+            // Have max no-progress-timeout quite high to be able to cope with huge
+            // I/O congestion spikes w/o failing in vain.
+            while ( !executor.awaitTermination( 30, SECONDS ) )
+            {
+                if ( lastProgression == nextId.get() )
+                {
+                    // No progression at all, abort?
+                    error.compareAndSet( null, new IOException( "No progress, so forcing abort" ) );
+                }
+                lastProgression = nextId.get();
+            }
+        }
+        catch ( InterruptedException e )
+        {
+            Thread.currentThread().interrupt();
+        }
+
+        Throwable finalError = error.get();
+        if ( finalError != null )
+        {
+            throw launderedException( IOException.class, finalError );
+        }
+
+        long endTime = currentTimeMillis();
+        monitor.recoveryCompleted( highTreeNodeId - lowTreeNodeId, cleanedPointers.get(), endTime - startTime );
+    }
+
+    private Runnable cleaner( AtomicLong nextId, AtomicReference<Throwable> error, AtomicInteger cleanedPointers )
+    {
+        return () ->
+        {
+            try ( PageCursor cursor = pagedFile.io( 0, PagedFile.PF_SHARED_READ_LOCK );
+                    PageCursor writeCursor = pagedFile.io( 0, PagedFile.PF_SHARED_WRITE_LOCK ) )
+            {
+                while ( nextId.get() < highTreeNodeId )
```

Comment:
it looks like if we will pass ranges if ids to check we can speed it up, since we will remove common contention point which is nextId currently, plus we will have ranges of ids in one thread that will potentially make it faster to process those because of locality